pkg/skaffold/deploy/docker/deploy.go

```diff
@@ -133,28 +145,90 @@ func (d *Deployer) deploy(ctx context.Context, out io.Writer, b graph.Artifact)
 		return fmt.Errorf("docker compose not yet supported by skaffold")
 	}
 
-	ports, bindings, err := d.portManager.getPorts(b.ImageName, d.resources)
+	containerCfg, err := d.containerConfigFromImage(ctx, artifact.Tag)
+	if err != nil {
+		return err
+	}
+
+	bindings, err := d.portManager.getPorts(artifact.ImageName, d.resources, containerCfg)
 	if err != nil {
 		return err
 	}
 
-	containerName := d.getContainerName(ctx, b.ImageName)
+	initContainers, err := d.debugger.TransformImage(ctx, artifact, containerCfg)
+	if err != nil {
+		return errors.Wrap(err, "transforming image for debugging")
+	}
```

Comment:
My point about `docker/debugger/transform.go`'s use of `transformImage` is essentially that the check should be done here.  So we're only calling TransformImage when we've enabled debugging.  It's just better to have this condition flow down here, as otherwise it reads like _we always transform images even when we're not debugging_.  And that's harder to reason about.
```suggestion
	var initContainers []*container.Config
	if d.debugger.shouldTransform {
		initContainers, err = d.debugger.TransformImage(ctx, artifact, containerCfg)
		if err != nil {
			return errors.Wrap(err, "transforming image for debugging")
		}
	}
```

I'd actually recommend moving `shouldTransform` into this package and rename it to `debuggingRequested` or something like that.

Alternatively you could move all the initContainer code below within this `if`.  We don't have any other usecases for init-containers.

Reply:
yeah that's a fair point. I changed this so that

1) if we're in `skaffold debug`, the `runDebug` function in `cmd/debug.go` sets `opts.Debug = true`
2) the `docker.Config` object has a `Debug()` method on it which pulls this boolean from the `opts`
3) if we're in debug mode, the `DebugManager` gets instantiated; otherwise, it's `nil`
4) during `Deploy()`, we check `if deployer.debugger == nil`, and conditionally execute debug-related logic on that check